File: puddle-twin-tokyo 2/backend/app/routing.py
"""歩行空間ネットワーク上の経路探索。

「距離だけを見る最短ルート」と「危険度をコストに織り込んだ回避ルート」を、
同じダイクストラ法にコスト関数だけ差し替えて求める。
"""

from __future__ import annotations

import heapq
import math
from dataclasses import dataclass

import numpy as np

from . import geo
from .config import (
    AVOID_ALPHA,
    EDGE_SAMPLE_STEP_M,
    EXPOSURE_REFERENCE_M,
    MAX_SNAP_DISTANCE_M,
    PUDDLE_REFERENCE_RISK,
    WALK_METERS_PER_MIN,
)
from .hazards import RiskModel
from .network import WalkGraph


@dataclass
class Route:
    """探索結果の1本ぶん。"""

    path: list[tuple[float, float]]  # [(lon, lat), ...]
    distance_m: float
    risk_score: int  # 0..100

    @property
    def duration_min(self) -> int:
        return max(1, round(self.distance_m / WALK_METERS_PER_MIN))


def annotate_edge_risk(graph: WalkGraph, model: RiskModel) -> None:
    """各辺に、その区間の平均危険度と最大危険度を書き込む。

    危険地点との距離ではなくラスタを直接サンプリングするので、
    「危険地点の代表点からは外れているが、区間全体がじわっと低い」道も拾える。
    """
    for edge in graph.edges:
        lon1, lat1 = graph.nodes[edge["a"]]
        lon2, lat2 = graph.nodes[edge["b"]]
        samples = [
            model.sample_risk(lon, lat)
            for lon, lat in geo.interpolate_points(
                lon1, lat1, lon2, lat2, step_m=EDGE_SAMPLE_STEP_M
            )
        ]
        edge["risk"] = round(float(np.mean(samples)), 4)
        edge["riskMax"] = round(float(np.max(samples)), 4)


def _effective_risk(edge: dict, rain_multiplier: float) -> float:
    """雨の強さを掛けた、その辺の実効危険度(0..1)。"""
    return min(1.0, edge.get("risk", 0.0) * rain_multiplier)


def _dijkstra(
    graph: WalkGraph, start: int, goal: int, rain_multiplier: float, avoid: bool
) -> list[int]:
    """start から goal までの最小コスト経路のノード列を返す。"""
    best_cost: dict[int, float] = {start: 0.0}
    previous: dict[int, int] = {}
    queue: list[tuple[float, int]] = [(0.0, start)]
    settled: set[int] = set()

    while queue:
        cost, node = heapq.heappop(queue)
        if node in settled:
            continue
        settled.add(node)
        if node == goal:
            break

        for neighbour, edge in graph.neighbours(node):
            if neighbour in settled:
                continue
            length = edge["length"]
            if avoid:
                # 危険な区間は「実際より長い道」として扱い、迂回を選ばせる
                step = length * (1.0 + AVOID_ALPHA * _effective_risk(edge, rain_multiplier))
            else:
                step = length
            candidate = cost + step
            if candidate < best_cost.get(neighbour, float("inf")):
                best_cost[neighbour] = candidate
                previous[neighbour] = node
                heapq.heappush(queue, (candidate, neighbour))

    if goal not in best_cost:
        return []

    node_path = [goal]
    while node_path[-1] != start:
        node_path.append(previous[node_path[-1]])
    node_path.reverse()
    return node_path


def _edges_along(graph: WalkGraph, node_path: list[int]) -> list[dict]:
    """ノード列から、実際に通る辺を取り出す。"""
    edges = []
    for a, b in zip(node_path, node_path[1:]):
        candidates = [
            edge
            for _, edge in graph.neighbours(a)
            if (edge["a"] == a and edge["b"] == b) or (edge["a"] == b and edge["b"] == a)
        ]
        if candidates:
            edges.append(min(candidates, key=lambda e: e["length"]))
    return edges


def score_route(edges: list[dict], rain_multiplier: float) -> int:
    """ルート全体の危険度を 0..100 で表す。

    「そのルートを歩いて水たまりに出くわす推定確率」として計算する。
    区間ごとの遭遇率を距離で積み上げ、1回も出くわさない確率の余事象を取る。
    危険度の高い区間を長く歩くほど上がり、短く抜ければ下がる。
    """
    if not edges:
        return 0

    exposure = 0.0
    for edge in edges:
        risk = _effective_risk(edge, rain_multiplier)
        encounter_rate = min(1.0, risk / PUDDLE_REFERENCE_RISK)
        exposure += encounter_rate * edge["length"] / EXPOSURE_REFERENCE_M

    probability = 1.0 - math.exp(-exposure)
    return int(round(min(1.0, max(0.0, probability)) * 100))


def find_routes(
    graph: WalkGraph,
    origin: tuple[float, float],
    destination: tuple[float, float],
    rain_multiplier: float,
) -> dict[str, Route]:
    """最短ルートと水たまり回避ルートをまとめて求める。

    origin / destination は (lon, lat)。いちばん近い歩行ネットワーク上の
    ノードへスナップしてから探索し、線の端はクリック地点まで伸ばす。
    """
    start, start_distance = graph.nearest_node_with_distance(*origin)
    goal, goal_distance = graph.nearest_node_with_distance(*destination)

    # 道から離れすぎた地点を無理に寄せると、ピンとルートの始点が
    # 大きく離れて見える。そうなる前に、選び直してもらう。
    for name, distance in (("出発地", start_distance), ("目的地", goal_distance)):
        if distance > MAX_SNAP_DISTANCE_M:
            raise ValueError(
                f"{name}の近くに歩ける道が見つかりません"
                f"(いちばん近い道まで約{distance:.0f}m)。道路の上で選び直してください"
            )

    if start == goal:
        raise ValueError("出発地と目的地が近すぎます。もう少し離れた地点を選んでください")

    routes: dict[str, Route] = {}
    for key, avoid in (("shortest", False), ("avoid", True)):
        node_path = _dijkstra(graph, start, goal, rain_multiplier, avoid=avoid)
        if not node_path:
            raise ValueError("出発地と目的地をつなぐ歩行ルートが見つかりませんでした")

        coords = [graph.nodes[node_id] for node_id in node_path]
        coords = [origin] + coords + [destination]
        edges = _edges_along(graph, node_path)
        routes[key] = Route(
            path=coords,
            distance_m=geo.path_length_m(coords),
            risk_score=score_route(edges, rain_multiplier),
        )

    return routes

File: frontend/src/engine/routing.ts
import { haversineMeters, pathLengthMeters, distanceToPath } from '../utils/geo';
import { nearestNode } from './graph';
import type { DangerPoint, LatLng, RainIntensity, RouteResult, SearchResult } from '../types';
import type { Edge, EngineData } from './data';

/**
 * ブラウザ側の経路探索。
 *
 * バックエンド(app/routing.py)と同じ手順を踏む。辺ごとの危険度と
 * 浸水フラグは walk_graph.json に焼き込み済みなので、地形の再計算は要らない。
 * 係数はすべて engine.json から取っている。
 */

/** 最小ヒープ。経路探索の待ち行列に使う。 */
class MinHeap {
  private items: { cost: number; node: number }[] = [];

  push(cost: number, node: number): void {
    this.items.push({ cost, node });
    let i = this.items.length - 1;
    while (i > 0) {
      const parent = (i - 1) >> 1;
      if (this.items[parent].cost <= this.items[i].cost) break;
      [this.items[parent], this.items[i]] = [this.items[i], this.items[parent]];
      i = parent;
    }
  }

  pop(): { cost: number; node: number } | undefined {
    const top = this.items[0];
    const last = this.items.pop();
    if (this.items.length > 0 && last) {
      this.items[0] = last;
      let i = 0;
      for (;;) {
        const left = i * 2 + 1;
        const right = left + 1;
        let smallest = i;
        if (left < this.items.length && this.items[left].cost < this.items[smallest].cost) smallest = left;
        if (right < this.items.length && this.items[right].cost < this.items[smallest].cost) smallest = right;
        if (smallest === i) break;
        [this.items[smallest], this.items[i]] = [this.items[i], this.items[smallest]];
        i = smallest;
      }
    }
    return top;
  }

  get size(): number {
    return this.items.length;
  }
}

function effectiveRisk(edge: Edge, rainMultiplier: number): number {
  return Math.min(1, edge.risk * rainMultiplier);
}

function dijkstra(
  data: EngineData,
  start: number,
  goal: number,
  rainMultiplier: number,
  avoid: boolean,
): number[] {
  const { graph, settings } = data;
  const best = new Map<number, number>([[start, 0]]);
  const previous = new Map<number, number>();
  const settled = new Set<number>();
  const queue = new MinHeap();
  queue.push(0, start);

  while (queue.size > 0) {
    const current = queue.pop()!;
    if (settled.has(current.node)) continue;
    settled.add(current.node);
    if (current.node === goal) break;

    for (const edgeIndex of graph.adjacency.get(current.node) ?? []) {
      const edge = graph.edges[edgeIndex];
      const next = edge.a === current.node ? edge.b : edge.a;
      if (settled.has(next)) continue;

      // 危険な区間は「実際より長い道」として扱い、迂回を選ばせる
      const step = avoid
        ? edge.length * (1 + settings.avoidAlpha * effectiveRisk(edge, rainMultiplier))
        : edge.length;
      const candidate = current.cost + step;
      if (candidate < (best.get(next) ?? Infinity)) {
        best.set(next, candidate);
        previous.set(next, current.node);
        queue.push(candidate, next);
      }
    }
  }

  if (!best.has(goal)) return [];
  const nodePath = [goal];
  while (nodePath[nodePath.length - 1] !== start) {
    const prior = previous.get(nodePath[nodePath.length - 1]);
    if (prior === undefined) return [];
    nodePath.push(prior);
  }
  return nodePath.reverse();
}

function edgesAlong(data: EngineData, nodePath: number[]): Edge[] {
  const { graph } = data;
  const edges: Edge[] = [];
  for (let i = 0; i + 1 < nodePath.length; i++) {
    const a = nodePath[i];
    const b = nodePath[i + 1];
    let chosen: Edge | null = null;
    for (const edgeIndex of graph.adjacency.get(a) ?? []) {
      const edge = graph.edges[edgeIndex];
      const connects = (edge.a === a && edge.b === b) || (edge.a === b && edge.b === a);
      if (connects && (chosen === null || edge.length < chosen.length)) chosen = edge;
    }
    if (chosen) edges.push(chosen);
  }
  return edges;
}

/**
 * ルート全体の危険度を 0..100 で表す。
 * そのルートを歩いて水たまりに出くわす推定確率として計算する。
 */
function scoreRoute(data: EngineData, edges: Edge[], rainMultiplier: number): number {
  const { settings } = data;
  if (edges.length === 0) return 0;

  let exposure = 0;
  for (const edge of edges) {
    const rate = Math.min(1, effectiveRisk(edge, rainMultiplier) / settings.puddleReferenceRisk);
    exposure += (rate * edge.length) / settings.exposureReferenceM;
  }
  return Math.round(Math.min(1, Math.max(0, 1 - Math.exp(-exposure))) * 100);
}

/** 東京都の浸水予想で浸水する区間が、経路に占める割合(%)。 */
function floodOverlap(edges: Edge[]): number {
  if (edges.length === 0) return 0;
  let total = 0;
  let flooded = 0;
  for (const edge of edges) {
    total += edge.length;
    flooded += edge.flood * edge.length;
  }
  return total <= 0 ? 0 : Math.round((flooded / total) * 100);
}

function levelFor(data: EngineData, weight: number): DangerPoint['level'] {
  if (weight >= data.settings.levelHighThreshold) return 'high';
  if (weight >= data.settings.levelMediumThreshold) return 'medium';
  return 'low';
}

function dangerPointsFor(
  data: EngineData,
  intensity: RainIntensity,
  paths: LatLng[][],
): DangerPoint[] {
  const { settings, hazards } = data;
  const multiplier = settings.rainProfiles[intensity].multiplier;

  const nearby = hazards.filter((hazard) => {
    const point = { lat: hazard.lat, lng: hazard.lng };
    return paths.some((path) => distanceToPath(point, path) <= settings.hazardCorridorM);
  });

  nearby.sort((a, b) => b.baseWeight - a.baseWeight);

  return nearby.slice(0, settings.hazardMaxResults).map((hazard) => {
    const weight = Math.min(1, hazard.baseWeight * multiplier);
    return {
      id: hazard.id,
      lat: hazard.lat,
      lng: hazard.lng,
      baseWeight: hazard.baseWeight,
      weight,
      displayRisk: Math.round(weight * 100),
      level: levelFor(data, weight),
      reason: hazard.reasons[intensity] ?? '',
    };
  });
}

export function computeRoutes(
  data: EngineData,
  origin: LatLng,
  destination: LatLng,
  intensity: RainIntensity,
): SearchResult {
  const { graph, settings } = data;
  const area = settings.demoArea;

  for (const [point, name] of [
    [origin, '出発地'],
    [destination, '目的地'],
  ] as const) {
    const inside =
      point.lat >= area.minLat &&
      point.lat <= area.maxLat &&
      point.lng >= area.minLng &&
      point.lng <= area.maxLng;
    if (!inside) {
      throw new Error(
        `${name}がデモ対象地域の外です。東京駅・有楽町駅・日比谷・京橋の周辺で指定してください`,
      );
    }
  }

  const start = nearestNode(graph, origin);
  const goal = nearestNode(graph, destination);

  for (const [snap, name] of [
    [start, '出発地'],
    [goal, '目的地'],
  ] as const) {
    if (snap.distanceM > settings.maxSnapDistanceM) {
      throw new Error(
        `${name}の近くに歩ける道が見つかりません（いちばん近い道まで約${snap.distanceM.toFixed(0)}m）。道路の上で選び直してください`,
      );
    }
  }

  if (start.id === goal.id) {
    throw new Error('出発地と目的地が近すぎます。もう少し離れた地点を選んでください');
  }

  const multiplier = settings.rainProfiles[intensity].multiplier;
  const routes: RouteResult[] = [];
  const paths: LatLng[][] = [];

  for (const id of ['shortest', 'avoid'] as const) {
    const nodePath = dijkstra(data, start.id, goal.id, multiplier, id === 'avoid');
    if (nodePath.length === 0) {
      throw new Error('出発地と目的地をつなぐ歩行ルートが見つかりませんでした');
    }

    const path = [origin, ...nodePath.map((n) => graph.nodes.get(n)!), destination];
    const edges = edgesAlong(data, nodePath);
    const distanceM = pathLengthMeters(path);

    paths.push(path);
    routes.push({
      id,
      label: settings.routeStyles[id].label,
      color: settings.routeStyles[id].color,
      distanceM: Math.round(distanceM),
      durationMin: Math.max(1, Math.round(distanceM / settings.walkMetersPerMin)),
      riskScore: scoreRoute(data, edges, multiplier),
      floodOverlapPct: floodOverlap(edges),
      path,
    });
  }

  return { routes, dangerPoints: dangerPointsFor(data, intensity, paths) };
}

export { haversineMeters };
